schema/schemaspec/extension.go

```diff
@@ -200,6 +200,30 @@ func (r *Resource) As(target interface{}) error {
 	return nil
 }
 
+// FinalName returns the final name for the resource by examining the struct tags for
+// the extension of the Resource's type. If no such extension is registered or the
+// extension struct does not have a name field, an error is returned.
+func (r *Resource) FinalName() (string, error) {
+	extensionsMu.RLock()
+	defer extensionsMu.RUnlock()
+	t, ok := extensions[r.Type]
```

Comment:
Feels not right to me to synchronize codec flow. Don't you have this info on the Resource itself?

Reply:
The lock protects concurrent reading and writing extensions (schemaspec.Register).
I did something similar to:
https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/database/sql/sql.go;l=817

Reply:
discussed offlien, will remove this in a later pr